impl/src/settings.rs

```diff
@@ -367,8 +380,95 @@ pub fn format_registry_url(registry_url: &str, name: &str, version: &str) -> Str
     .replace("{version}", version)
 }
 
+fn validate_crate_setting_additional_build_file_is_relative(
+  file_path: &Path,
+  crate_name: &str,
+  version: &VersionReq,
+) -> Result<(), RazeError> {
+  if file_path.is_absolute() {
+    return Err(RazeError::Config {
+      field_path_opt: Some(format!(
+        "raze.crates.{}.{}.additional_build_file",
+        crate_name,
+        version.to_string()
+      )),
+      message: "additional_build_file flags should be relative paths from the manifest containing \
+                the definition"
+        .to_owned(),
+    });
+  }
+
+  Ok(())
+}
+
+/** Ensures crate settings which specify additional build files have canonicalized paths */
+fn canonicalize_crate_setting_additional_build_file(
+  additional_build_file: &mut &mut PathBuf,
+  toml_path: &Path,
+  crate_name: &str,
+  version: &VersionReq,
+) -> Result<()> {
+  validate_crate_setting_additional_build_file_is_relative(
+    additional_build_file,
+    crate_name,
+    version,
+  )?;
+
+  **additional_build_file = toml_path.join(&additional_build_file).canonicalize()?;
+
+  Ok(())
+}
+
+/** Ensures crate settings associatd with the parsed [RazeSettings](crate::settings::RazeSettings) have valid crate settings */
+fn validate_crate_settings(settings: &mut RazeSettings, toml_path: &Path) -> Result<(), RazeError> {
+  let mut errors = Vec::new();
+
+  for (crate_name, crate_settings) in settings.crates.iter_mut() {
+    for (version, crate_settings) in crate_settings.iter_mut() {
+      if crate_settings.additional_build_file.is_none() {
+        continue;
+      }
+
+      let additional_build_file: &mut &mut PathBuf =
+        &mut crate_settings.additional_build_file.as_mut().unwrap();
+
+      let result = canonicalize_crate_setting_additional_build_file(
+        additional_build_file,
+        toml_path,
+        crate_name,
+        version,
+      );
+
+      if result.is_err() {
+        errors.push(RazeError::Config {
+          field_path_opt: Some(format!(
+            "raze.crates.{}.{}.additional_build_file",
+            crate_name,
+            version.to_string()
+          )),
+          message: format!(
+            "Failed to canonicalize string with error: {}",
+            // UNWRAP: safe inside this block
+            result.err().unwrap().to_string()
+          ),
+        });
+      }
+    }
+  }
+
+  // Surface all errors
+  if !errors.is_empty() {
+    return Err(RazeError::Config {
+      field_path_opt: None,
+      message: format!("{:?}", errors),
+    });
+  }
+
+  Ok(())
+}
+
 /** Verifies that the provided settings make sense. */
-fn validate_settings(settings: &mut RazeSettings) -> Result<(), RazeError> {
+fn validate_settings(settings: &mut RazeSettings, toml_path: &Path) -> Result<(), RazeError> {
```

Comment:
"toml_path" here means "workspace_root_toml". Actually, not even that, it means "workspace_root_toml_directory".

Reply:
It only means that in this context but it's incorrect to assume it will always be the root toml path

Reply:
I dug in a bit more here, and I think I'm right here. Conceptually, yes, this could be an arbitrary toml path. But in practice, this is two things. First, it's a directory which contains a toml (a minor quible). Second, this is only called one time, in the context of the root (workspace) toml, after the concatenative shenanigans of parse_raze_settings and parse workspace settings and all of that.

Reply:
I updated this. I think it's better now